File: Week_01/917.仅仅反转字母.c
/*
 * @lc app=leetcode.cn id=917 lang=c
 *
 * [917] 仅仅反转字母
 */

// @lc code=start


char * reverseOnlyLetters(char * S){
    char *left = S;
    char *right = S + strlen(S) - 1;
    while(left < right)
    {
        if(isalpha(*left) && isalpha(*right))
        {
            char temp = *left;
            *left = *right;
            *right = temp;
            left++;
            right--;
        }
        else if(isalpha(*left))
        {
            right--;
        }
        else
        {
            left++;
        }
    }
    return S;
}


// @lc code=end

File: Week_01/917.仅仅反转字母.java
class Solution {
    public String reverseOnlyLetters(String S) {
        int left = 0;
        int right = S.length();
        StringBuilder sb = new StringBuilder(S);
        while (left < right) {
            if (Character.isLetter(S.charAt(left)) && Character.isLetter(S.charAt(right))) {
                sb.setCharAt(left, S.charAt(right));
                sb.setCharAt(right, S.charAt(left));
                left++;
                right--;
            } else if (Character.isLetter(S.charAt(left))) {
                right--;
            } else {
                left++;
            }
        }
        return sb.toString();
    }
}